ch12.md

```diff
@@ -0,0 +1,256 @@
+
+## No law and order
+
+Well now, before you get all judgemental and bang the backspace button like a gavel to retreat from the chapter, take a moment to recognize that these laws are useful code guarantees. Tis' my conjecture that the goal of most program architecture is an attempt to place useful restrictions on our code to narrow the possibilities, to guide us into the answers as designers and readers.
+
+An interface without laws is merely indirection. Like any other mathematical structure, we must expose properties for our own sanity. This has a similar effect as encapsulation since it protects the data, enabling us to swap out the interface with another law abiding citizen.
+
+Come along now, we've got some laws to suss out.
+
+### Identity
+
+```js
+const identity1 = compose(sequence(Identity.of), map(Identity.of))
+const identity2 = Identity.of
+
+// test it out with Right
+identity1(Right('stuff'))
+// Identity(Right('stuff'))
+
+identity2(Right('stuff'))
+// Identity(Right('stuff'))
+```
+
+This should be straight forward. If we place an `Identity` in our functor, then turn it inside out with `sequence` that's the same as just placing it on the outside to begin with. We chose `Right` as our guinea pig as it is easy to try the law and inspect. An arbitrary functor there is normal, however, the use of a concrete functor here, namely `Identity` in the law itself might raise some eyebrows. Remember a [category](ch5.md#category-theory) is defined by morphisms between its objects that have associative composition and identity. When dealing with the category of functors, natural transformations are the morphisms and `Identity` is, well identity. The `Identity` functor is as fundamental in demonstrating laws as our `compose` function. In fact, we should give up the ghost and follow suit with our [Compose](ch8.md#a-spot-of-theory) type:
+
+### Composition
+
+```js
+const comp1 = compose(sequence(Compose.of), map(Compose.of))
+const comp2 = (point1, point2) => compose(Compose.of, map(sequence(point2)), sequence(point1))
+
+
+// Test it out with some types we have lying around
+comp1(Identity(Right([true])))
+// Compose(Right([Id(true)]))
+
+comp2(Right, Array)(Identity(Right([true])))
+// Compose(Right([Id(true)]))
+```
+
+This law preserves composition as one would expect: if we swap compositions of functors, we shouldn't see any surprises since the composition is a functor itself. We arbitrarily chose `true`, `Right`, `Identity`, and `Array` to test it out. Libraries like [quickcheck](https://hackage.haskell.org/package/QuickCheck) or [jsverify](http://jsverify.github.io/) can help us test the law by fuzz testing the inputs.
+
+As a natural consequence of the above law, we get the ability to fuse traversals:
+
+```js
+traverse(compose(Compose.of, map(compose(f , g)))) = compose(Compose.of, map(traverse(of, f)), traverse(of, g))
```

Comment:
Indeed. Plus, I believe the law is erroneous. Let's switch to `PureScript` to lose some verbosity and gain some expressiveness, we currently have this:

```purs
traverse (Compose . map (f . g)) = Compose . map (traverse f) . traverse g
```

On the left, it suggests that:

```purs
g :: a -> b
f :: b -> f c
Compose :: h (f c) -> Compose h f c
```

and on the right:

```purs
g :: a -> g b
f :: b -> f c
Compose :: g (f c) -> Compose g f c
```

which doesn't hold well. I believe it would rather be:

```purs
traverse (Compose . map f . g) = Compose . map (traverse f) . traverse g

-- or even, to have (Compose f g) instead of (Compose g f)

traverse (Compose . map g . f) = Compose . map (traverse g) . traverse f
```

and, translated with the js version:

```js
traverse(Compose.of, compose(Compose.of, map(g), f)) === compose(Compose.of, map(traverse(of$2, g)), traverse(of$1, f))

// where:
// of$1 :: a -> f a
// of$2 :: b -> g b
```

@DrBoolean, can you confirm ?

Reply:
Ahh, yes, sorry. That one was never verified, just loosely translated in a hurry. Nice work on the Purs proof ;)

Reply:
So i translated the fusion law from https://www.cs.ox.ac.uk/jeremy.gibbons/publications/iterator.pdf

NO CLUE WHERE I GOT THE `map`?! wtf.

Reply:
Just provided a link here instead of writing it out.

Reply:
Indeed:

![image](https://user-images.githubusercontent.com/5680256/34457276-0aa2b98e-edac-11e7-98f9-cd69ff298497.png)

Also confirmed with [hackage](http://hackage.haskell.org/package/base-4.10.0.0/docs/Data-Traversable.html#t:Traversable)'s description of Traversable ^.^